// @cra-express/static-loader/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var express = require("express");

var bodyParser = require("body-parser");

var compression = require("compression");

function staticLoader(app, options) {
  var clientBuildPath = options.clientBuildPath;
  app.use(compression());
  app.use(bodyParser.json());
  app.use(bodyParser.urlencoded({
    extended: false
  })); // Serve static assets

  if (process.env.NODE_ENV === "development") {
    // Connect proxy to Create React App dev server
    var proxy = require("http-proxy-middleware");

    var craServiceName = process.env.CRA_SERVICE_NAME || 'localhost';
    var craClientPort = process.env.CRA_CLIENT_PORT || 3000;
    app.use(["**/*.*", "/static", "/sockjs-node"], proxy({
      target: "http://".concat(craServiceName, ":").concat(craClientPort),
      changeOrigin: true,
      ws: true
    }));
    console.log("Connected to CRA Client dev server");
  } else {
    app.use(express.static(clientBuildPath, {
      index: false
    }));
  }

  return app;
}

var _default = staticLoader;
exports.default = _default;
